File: I1_MVPAndRetrofit/app/src/main/java/com/example/tsvetelinastoyanova/weatherreportrevisited/data/source/local/LocalDataSource.java
package com.example.tsvetelinastoyanova.weatherreportrevisited.data.source.local;

import android.support.annotation.NonNull;

import com.example.tsvetelinastoyanova.weatherreportrevisited.data.CityEntity;
import com.example.tsvetelinastoyanova.weatherreportrevisited.data.source.CityDataSource;

import java.util.List;

public interface LocalDataSource extends CityDataSource {
    interface LoadCitiesCallback {

        void onCitiesLoaded(List<CityEntity> cities);

        void onDataNotAvailable();

    }

    interface AddCityCallback{

        void onCityAddedSuccessfully(CityEntity cityEntity);

        void onFail();
    }

    interface DeleteCityCallback{

        void onCityDeletedSuccessfully();

        void onFail();
    }

    interface RefreshCityCallback{

        void onRefreshCitySuccessfully();

    }

    void getCities(@NonNull LoadCitiesCallback callback);

    void addCity(@NonNull CityEntity cityEntity, AddCityCallback addCityCallback);

    void refreshCity(@NonNull CityEntity cityEntity, LocalDataSource.RefreshCityCallback refreshCityCallback);

    void deleteCity(@NonNull String cityName, LocalDataSource.DeleteCityCallback callback);
}

File: I1_MVPAndRetrofit_Kotlin/app/src/main/java/com/example/tsvetelinastoyanova/weatherapp/data/source/local/LocalDataSource.kt
package com.example.tsvetelinastoyanova.weatherapp.data.source.local

import com.example.tsvetelinastoyanova.weatherapp.data.CityEntity
import com.example.tsvetelinastoyanova.weatherapp.data.source.CityDataSource
import io.reactivex.Flowable
import io.reactivex.Single


interface LocalDataSource : CityDataSource {

    interface AddCityCallback {

        fun onCityAddedSuccessfully(cityEntity: CityEntity)

        fun onCityExistsInDatabase()

        fun onNotValidCity()
    }

    interface DeleteCityCallback {

        fun onCityDeletedSuccessfully()

        fun onFail()
    }

    fun getCities(): Single<List<CityEntity>>

    fun addCity(cityEntity: CityEntity): Single<CityEntity>

    fun refreshCity(cityEntity: CityEntity): Single<CityEntity>

    fun deleteCity(cityName: String): Single<String>
}
